Update to latest nixpkgs-unstable, rustc

mmilata opened the pull request:
<!--
If you are a core dev:

Don't forget to set up the fields:
- assign yourself
- set priority to same as original issue
- add PR to sprint
- if Draft -> "in progress"
- if final PR -> "needs review"

If you're an external contributor, you can ignore this text.
-->

cepetr commented:
When you run `poetry install` on macOS you can see this:
```
Consider moving TOML configuration files to /Users/{user}/Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
```

and this: 

```
Warning: The current project could not be installed: No file/folder found for package trezor-firmware
If you do not want to install the current project use --no-root.
If you want to use Poetry only for dependency management but not for packaging, you can disable package mode by setting package-mode = false in your pyproject.toml file.
In a future version of Poetry this warning will become an error!
```

(maybe these're not new)

mmilata commented:
Still, even with the micropython fix device tests crash on something. I'm giving up for now.

```
$ make test_emu
/Users/mm/work/trezor-firmware/core/emu.py --disable-animation --headless --output=/Users/mm/work/trezor-firmware/core/../tests/trezor.log --temporary-profile -c pytest --junitxml=/Users/mm/work/trezor-firmware/core/../tests/junit.xml /Users/mm/work/trezor-firmware/core/../tests/device_tests  --lang="en"
TREZOR_PATH=udp:127.0.0.1:21324
TREZOR_PROFILE_DIR=/private/tmp/nix-shell-45162-0/trezor-emulator-5esd7p_9
TREZOR_UDP_PORT=21324
TREZOR_FIDO2_UDP_PORT=21326
TREZOR_SRC=/Users/mm/work/trezor-firmware/core/src
Waiting for emulator to come up...
Emulator ready after 0.417 seconds
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.10.10, pytest-8.3.2, pluggy-1.5.0
Using --random-order-bucket=module
Using --random-order-seed=389422

rootdir: /Users/mm/work/trezor-firmware
configfile: setup.cfg
plugins: xdist-2.5.0, forked-1.4.0, random-order-1.0.4, flaky-3.8.1, timeout-2.1.0, hypothesis-6.36.1, ordering-0.6
collecting ... Fatal Python error: Aborted

Current thread 0x00000002022af840 (most recent call first):
  File "/nix/store/r23sw9iy2jdrh59dsq8mic6ahkcijijr-python3-3.10.10/lib/python3.10/ctypes/__init__.py", line 374 in __init__
  File "/nix/store/r23sw9iy2jdrh59dsq8mic6ahkcijijr-python3-3.10.10/lib/python3.10/ctypes/__init__.py", line 452 in LoadLibrary
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/bitcoin/core/key.py", line 34 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 883 in exec_module
  File "<frozen importlib._bootstrap>", line 688 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1006 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1027 in _find_and_load
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/bitcoin/wallet.py", line 33 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 883 in exec_module
  File "<frozen importlib._bootstrap>", line 688 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1006 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1027 in _find_and_load
  File "/Users/martinmilata/work/trezor-firmware/tests/device_tests/bitcoin/signtx.py", line 8 in <module>
  File "<frozen importlib._bootstrap>", line 241 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 883 in exec_module
  File "<frozen importlib._bootstrap>", line 688 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1006 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1027 in _find_and_load
  File "/Users/martinmilata/work/trezor-firmware/tests/device_tests/bitcoin/test_authorize_coinjoin.py", line 29 in <module>
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/assertion/rewrite.py", line 174 in exec_module
  File "<frozen importlib._bootstrap>", line 688 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1006 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1027 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1050 in _gcd_import
  File "/nix/store/r23sw9iy2jdrh59dsq8mic6ahkcijijr-python3-3.10.10/lib/python3.10/importlib/__init__.py", line 126 in import_module
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/pathlib.py", line 582 in import_path
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/python.py", line 493 in importtestmodule
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/python.py", line 546 in _getobj
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/python.py", line 284 in obj
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/python.py", line 562 in _register_setup_module_fixture
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/python.py", line 549 in collect
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/runner.py", line 389 in collect
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/runner.py", line 341 in from_call
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/runner.py", line 391 in pytest_make_collect_report
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/runner.py", line 567 in collect_one_node
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 835 in _collect_one_node
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 970 in genitems
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 975 in genitems
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 975 in genitems
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 809 in perform_collect
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 347 in pytest_collection
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 336 in _main
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/Users/martinmilata/work/trezor-firmware/.venv/lib/python3.10/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/Users/martinmilata/work/trezor-firmware/.venv/bin/pytest", line 8 in <module>

Extension modules: PIL._imaging (total: 1)
Terminating emulator...
Emulator shut down after 0.001 seconds
make: *** [Makefile:191: test_emu] Error 250
(trezor-firmware-py3.10) Mac:core martinmilata$
```

mmilata commented:
Ok, this seems to happen on `main` as well and can be considered out of scope for the PR.

The message `WARNING: python is loading libcrypto in an unsafe way` preceding the abort() leads to some clues but most of the solutions involve linking/copying libraries to /usr/local/lib which seems deranged and also doesn't work for me.

mmilata commented:
> Looks like a micropython bug

Turns out I can't read and it's probably a GCC problem. Need to try again with GCC 14 or a clang version from nixpkgs.

onvej-sl commented:
The [failing valgrind crypto test](https://github.com/trezor/trezor-firmware/actions/runs/12049602602/job/33596957680?pr=4374) appears similar to what we dealt with [here](https://github.com/trezor/trezor-firmware/pull/4146#issuecomment-2348857197). I also discovered this is caused by updating `gcc12`  to `gcc14`. I will look into it later.

mmilata commented:
Yep, documented what I found in #4393. Not sure how useful is testing this on `x86_64` but it probably can catch some cases which are not dependent on the compiler like this one. For the time being I think we can use GCC 14 on macOS only and stay on 12 on Linux (cross compiler is currently GCC 12, with 13 available).

mmilata commented:
Seems to be in somewhat workable state now.

@cepetr when you have time can you please test on macOS and review the micropython changes?

mmilata commented:
* `ERROR: Assert FILE: qstr.c:133` macOS crash is not fixed by GCC14, i just screwed up the testing, it needs PYOPT=0 to manifest
* `SDL2` update somehow breaks HIGHDPI support on macOS making the emulator unusable

cepetr commented:
> Seems to be in somewhat workable state now.
> 
> @cepetr when you have time can you please test on macOS and review the micropython changes?

`nix-shell --arg devTools true` seems to be broken again:

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating derivation 'trezor-firmware-env'
         whose name attribute is located at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'trezor-firmware-env'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/pkgs/stdenv/generic/make-derivation.nix:383:7:
          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Package ‘kcachegrind-23.08.5’ in /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/pkgs/applications/kde/kcachegrind.nix:12 is not available on the requested hostPlatform:
         hostPlatform.config = "aarch64-apple-darwin"
         package.meta.platforms = [
           "aarch64-linux"
           "armv5tel-linux"
           "armv6l-linux"
           "armv7a-linux"
           "armv7l-linux"
           "i686-linux"
           "loongarch64-linux"
           "m68k-linux"
           "microblaze-linux"
           "microblazeel-linux"
           "mips-linux"
           "mips64-linux"
           "mips64el-linux"
           "mipsel-linux"
           "powerpc64-linux"
           "powerpc64le-linux"
           "riscv32-linux"
           "riscv64-linux"
           "s390-linux"
           "s390x-linux"
           "x86_64-linux"
         ]
         package.meta.badPlatforms = [ ]
       , refusing to evaluate.
```

mmilata commented:
> nix-shell --arg devTools true seems to be broken again

Fixed in [311906c](https://github.com/trezor/trezor-firmware/pull/4374/commits/311906c62e899d9ba6327d65343abcf2a38c5871), please test when you have spare cycles.

Also brought back micropython/micropython#16300 and added questionable fix for emulator image scaling on macOS.

cepetr commented:
> > nix-shell --arg devTools true seems to be broken again
> 
> Fixed in [311906c](https://github.com/trezor/trezor-firmware/pull/4374/commits/311906c62e899d9ba6327d65343abcf2a38c5871), please test when you have spare cycles.

It seems, it's ok now. Thanks.

cepetr commented:
> > Consider moving TOML configuration files to /Users/{user}/Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
> 
> I can't see this one, perhaps I don't have anything in the legacy location wherever it is?

Looks like the problem is specific to my machine. I fixed it by manually moving `config.toml` to the expected folder.

prusnak commented on shell.nix:
```diff
@@ -121,27 +93,31 @@ stdenvNoCC.mkDerivation ({
+    # GCC 14 causes crypto tests to fail in CI due to emitting non-constant-time instructions,
+    # and it's probably a good idea to keep it the same version as gcc-arm-embedded anyway
+    # https://github.com/trezor/trezor-firmware/issues/4393
+    (if stdenv.isDarwin then gcc14 else gcc12)
```
can't we keep gcc12 for Darwin too?

prusnak commented:
While we are at it we should replace `gcc-arm-embedded` with `pkgsCross.arm-embedded-nano.buildPackages.gcc` and 
`gcc-arm-embedded-gdbfix` with `pkgsCross.arm-embedded-nano.buildPackages.gdb`

prusnak commented:
Btw, I recently updated gcc-arm-embedded to version 14 in nixpkgs master. We might want to check if it brings some benefits (let's say smaller binaries) or some build failures that could eventually mean fixing some bugs.

cepetr commented:
> @cepetr can you please try again when you have time? Make sure to update submodules, otherwise the emulator crashes on start.

I tested it on my machine, and debugging debugging seem to work as expected. The only issue I see is the unsupported `-g3` option which causes hundreds of warnings when building the emulator. I think we can replace it with just `-g` for darwin (or maybe for all platforms)

mmilata commented:
Switched to `-g` in 4eeb47f, let's see if it breaks anyone's debugging setup.

mmilata commented on core/SConscript.boardloader:
```diff
@@ -107,7 +107,7 @@ env.Replace(
 env.Replace(
     COPT=env.get('ENV').get('OPTIMIZE', '-Os'),
     CCFLAGS='$COPT '
-    '-g3 '
```
Oops ... fixed in e8c940a

cepetr reviewed (approved):
From my side (build on macos, hw debugging), it looks good.

mmilata commented:
Merged, thanks @cepetr!

romanz commented:
IIUC, ASAN tests are failing on nightly CI, starting from a98a6b0026:
https://github.com/trezor/trezor-firmware/actions/runs/14504487753/job/40691430263
The error can be reproduced locally using:
```
$ nix-shell --run "poetry shell"
$ poetry install --sync
$ ADDRESS_SANITIZER=1 TREZOR_MODEL=T2T1 PYOPT=0 QUIET_MODE=1 make -C core build_unix_frozen
$ core/emu.py -sea -c true
TREZOR_PATH=udp:127.0.0.1:21324
TREZOR_PROFILE_DIR=/var/tmp
TREZOR_UDP_PORT=21324
TREZOR_FIDO2_UDP_PORT=21326
TREZOR_SRC=/home/rzeyde/src/trezor-firmware/core/src
Waiting for emulator to come up...
FATAL: uncaught NLR 0x78902b83d8e0

=================================================================
==1086808==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7890316fc7d7 in malloc (/nix/store/7hyxhin05law4mnxsdrmqv30h0rnqdxm-gcc-14.2.0-lib/lib/libasan.so.8+0xfc7d7)
    #1 0x7890312cf4b2 in SDL_malloc_REAL (/nix/store/6a4s1c1d3xcl0i9lacj23ziz8l27zn5b-SDL2-2.30.6/lib/libSDL2-2.0.so.0+0xcf4b2)

SUMMARY: AddressSanitizer: 72 byte(s) leaked in 1 allocation(s).
Traceback (most recent call last):
  File "/home/rzeyde/src/trezor-firmware/core/emu.py", line 317, in <module>
    cli()
  File "/home/rzeyde/src/trezor-firmware/.venv/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzeyde/src/trezor-firmware/.venv/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/rzeyde/src/trezor-firmware/.venv/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzeyde/src/trezor-firmware/.venv/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzeyde/src/trezor-firmware/core/emu.py", line 278, in cli
    emulator.start()
  File "/home/rzeyde/src/trezor-firmware/python/src/trezorlib/_internal/emulator.py", line 180, in start
    self.wait_until_ready()
  File "/home/rzeyde/src/trezor-firmware/python/src/trezorlib/_internal/emulator.py", line 128, in wait_until_ready
    raise RuntimeError("Emulator process died")
RuntimeError: Emulator process died
```

romanz commented:
<details>

Using the diff below, we get the following NLR debug log:
```
Waiting for emulator to come up...
nlr_push: before: nlr_top=(nil), val=0x725ac4309ca0
nlr_push: before: nlr_top=0x725ac4309ca0, val=0x725ac4209360
nlr_raise: nlr_top=0x725ac4209360
nlr_raise: nlr_top=(nil)
FATAL: uncaught NLR 0x725ac143d920
```

<summary> SUMMARY </summary>

```diff
diff --git a/py/nlr.h b/py/nlr.h
index dade10771..d8eb0e2df 100644
--- a/py/nlr.h
+++ b/py/nlr.h
@@ -151,14 +151,15 @@ NORETURN void nlr_jump(void *val);
 NORETURN void nlr_jump_fail(void *val);
 
 // use nlr_raise instead of nlr_jump so that debugging is easier
-#ifndef MICROPY_DEBUG_NLR
+#if 0
 #define nlr_raise(val) nlr_jump(MP_OBJ_TO_PTR(val))
 #else
 #include "mpstate.h"
+#include <stdio.h>
 #define nlr_raise(val) \
     do { \
-        /*printf("nlr_raise: nlr_top=%p\n", MP_STATE_THREAD(nlr_top)); \
-        fflush(stdout);*/ \
+        printf("nlr_raise: nlr_top=%p\n", MP_STATE_THREAD(nlr_top)); \
+        fflush(stdout); \
         void *_val = MP_OBJ_TO_PTR(val); \
         assert(_val != NULL); \
         assert(mp_obj_is_exception_instance(val)); \
@@ -166,13 +167,8 @@ NORETURN void nlr_jump_fail(void *val);
     } while (0)
 
 #if !MICROPY_NLR_SETJMP
-#define nlr_push(val) \
-    assert(MP_STATE_THREAD(nlr_top) != val), nlr_push(val)
-
-/*
 #define nlr_push(val) \
     printf("nlr_push: before: nlr_top=%p, val=%p\n", MP_STATE_THREAD(nlr_top), val),assert(MP_STATE_THREAD(nlr_top) != val),nlr_push(val)
-*/
 #endif
 
 #endif
```

</details>

romanz commented:
Using GDB, we get the following backtrace:
```sh
$ core/emu.py -ea -D
<snip>
(gdb) b nlr_jump_fail
Breakpoint 1 at 0x5f607d: file embed/projects/unix/main.c, line 795.
(gdb) r
Starting program: ./core/build/unix/trezor-emu-core -O0 -X heapsize=20M -m main

Breakpoint 1, nlr_jump_fail (val=val@entry=0x7ffff183d8e0) at embed/projects/unix/main.c:795
795	  printf("FATAL: uncaught NLR %p\n", val);
(gdb) bt
#0  nlr_jump_fail (val=val@entry=0x7ffff183d8e0) at embed/projects/unix/main.c:795
#1  0x00000000005776e9 in nlr_jump (val=val@entry=0x7ffff183d8e0) at vendor/micropython/py/nlrx64.c:86
#2  0x00000000005c940b in mp_raise_type_arg (exc_type=0x95ace0 <mp_type_RuntimeError>, arg=arg@entry=0x449a) at vendor/micropython/py/runtime.c:1650
#3  0x00000000005c947d in mp_raise_recursion_depth () at vendor/micropython/py/runtime.c:1667
#4  0x00000000005cf225 in mp_stack_check () at vendor/micropython/py/stackctrl.c:60
#5  0x00000000005779b3 in mp_obj_print_helper (print=print@entry=0x8c2360 <mp_stderr_print>, o_in=o_in@entry=0x7ffff183d840, kind=kind@entry=PRINT_EXC) at vendor/micropython/py/obj.c:111
#6  0x0000000000577e39 in mp_obj_print_exception (print=print@entry=0x8c2360 <mp_stderr_print>, exc=exc@entry=0x7ffff183d840) at vendor/micropython/py/obj.c:154
#7  0x00000000005f3a37 in handle_uncaught_exception (exc=0x7ffff183d840) at embed/projects/unix/main.c:111
#8  0x00000000005f3d13 in do_import_module (modname=0x7fffffff84d5 "main") at embed/projects/unix/main.c:469
#9  0x00000000005f573d in main_ (argc=argc@entry=6, argv=argv@entry=0x7fffffff7da8) at embed/projects/unix/main.c:660
#10 0x0000000000407966 in main (argc=6, argv=0x7fffffff7da8) at embed/projects/unix/main_main.c:30
```

mmilata commented:
My current theory is that `MP_STATE_THREAD(stack_top)` (aka `stack_top`) gets set to a bad value. Later this gets compared to current stack frame and the resulting stack size is negative (i.e. near 2^64 when taken as unsigned, as seen in @romanz's last post). This makes micropython raise the recursion depth exception but we are already handling another exception in the outermost handler so it terminates with `uncaught NLR`.

To examine the original exception I disabled the stack usage check. It seems to come from this [gc.collect()](https://github.com/trezor/trezor-firmware/blob/9037c398eb47aa5f5230b076e3ec3e62d8deb1dd/core/src/storage/cache.py#L13) call while loading main.py. On the micropython side, again bad stack size is calculated and passed to `gc_collect_root` @ [shared/runtime/gchelper_generic.c](https://github.com/trezor/micropython/blob/476c569f3512f634b7f2e356d754bc17e5348d10/shared/runtime/gchelper_generic.c#L180).
